File: 20180715/Shortest Unsorted Continuous Subarray.java
class Solution {
    public int findUnsortedSubarray(int[] nums) {
        int l = 0, j = -1;
        int r = nums.length - 1, i = 0;
        int min_nums = Integer.MAX_VALUE, max_nums = Integer.MIN_VALUE;
        while( r >= 0){
            max_nums = Math.max(max_nums, nums[l]);
            if (nums[l] != max_nums){
                j = l;
            }
            min_nums = Math.min(min_nums, nums[r]);
            if(nums[r] != min_nums){
                i = r;
            }
            r -= 1;
            l += 1;
        }
        return j - i + 1;

    }
    
}

File: 20180715/Shortest Unsorted Continuous Subarray.py
class Solution(object):
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        i = 0
        j = len(nums) - 1
        while i < len(nums) - 1 and nums[i] <= nums[i+1]:
            i += 1
        while j > 0 and nums[j] >= nums[j-1]:
            j -= 1
        if (i == len(nums) - 1 and j == 0):
            return 0
        temp = nums[i:j+1]
        min_temp = min(temp)
        max_temp = max(temp)
        while i >= 1 and nums[i-1] > min_temp:
            i -= 1
        while j <= len(nums) - 2 and nums[j + 1] < max_temp:
            j += 1
        return j - i + 1
        
        
